rescript/replace_taxonomy.py

```diff
@@ -0,0 +1,50 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2021, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+
+from qiime2.plugin import MetadataColumn, Categorical
+import pandas as pd
+import re
+
+
+def make_substitutions_dict(taxonomy_replacement_map):
+    replacement_dict = taxonomy_replacement_map.to_series().to_dict()
+    return replacement_dict
+
+
+def make_regex(substitutions):
+    strings = substitutions.keys()
+    regex = re.compile('|'.join(map(re.escape, strings)))
+    return regex
+
+
+def lineages_with_invalid_number_of_ranks(updated_tax_series,
+                                          num_expected_ranks):
+    invalid_lineages = [label for label, lineage in updated_tax_series.items()
+                        if len(lineage.split(';')) != num_expected_ranks]
+    return invalid_lineages
+
+
+def replace_taxonomy(taxonomy: pd.Series,
+                     taxonomy_replacement_map: MetadataColumn[Categorical],
+                     num_expected_ranks: int = 7,
+                     ) -> pd.Series:
+
+    rd = make_substitutions_dict(taxonomy_replacement_map)
+    mr = make_regex(rd)
+
+    updated_tax_series = taxonomy.apply(lambda taxonomy:
+                                        mr.sub(lambda match:
+                                               rd[match.group(0)], taxonomy))
+
+    irl = lineages_with_invalid_number_of_ranks(updated_tax_series,
```

Comment:
is this just a safety catch to make sure that e.g., a greedy regex does not blow up the taxonomy?

there might be cases where this does not work, e.g., in many cases this function might be applied to taxonomic classifications of unknown sequences, which might have uneven numbers of ranks. So at the very least we should allow users to disable this function and place it in an `if` statement (`if num_expected_ranks:...`)

Do you have a good example of when this option is needed? if not, maybe it is unnecessary at this stage.

Another way to approach this more dynamically could be to count the number of ranks before and after replacement, instead of specifying a fixed length, like so:
```
ranks_in = taxonomy.str.count(';')
ranks_out = updated_tax_series.str.count(';')
if not ranks_in.compare(ranks_out).empty:
    raise ValueError("blah blah blah")
```

but again unless if you have a specific use case I wonder if this check is even needed (or could be moved to a separate validation function, e.g., to check that a taxonomy has even ranks)

Reply:
I just imagined many users might produce invalid taxonomy files. Which they won't be aware of until they proceed to downstream steps.

I came across this myself, when the best way to find a pattern was to include the `;` in the search string... and then I forgot to put it back with the replace string. Then I was missing a rank-level for that lineage.

I like your dynamic idea, particularly with the classification output! I'll work on implementing this route.

Reply:
oooh very good example... I agree, some output validation is in order (to check for empty strings if not equal lengths)... I think the "dynamic" approach should scratch that itch...

one remaining question is whether we want to allow users to bypass this check, e.g., if they intentionally remove a rank for some reason... I am not sure when/why someone would want to, but something to chew on.

Reply:
I have implemented a `warnings.warn`... But that'll only come up with the user runs `--verbose`. I can probably to a `try-except-finally`, so that it'll appear regardless of the `--verbose` flag being used?